api/v1alpha1/shared_types.go

```diff
@@ -169,3 +170,29 @@ func meta(datacenter string) map[string]string {
 		common.DatacenterKey: datacenter,
 	}
 }
+
+// transparentProxyConfigComparer compares two TransparentProxyConfig pointers.
+// It returns whether they are equal but will treat an empty struct and a nil
+// pointer as equal. This is needed to fix a bug in the Consul API in Consul
+// 1.10.0 (https://github.com/hashicorp/consul/issues/10595) where Consul will
+// always return the empty struct for the TransparentProxy key even if it was
+// written as a nil pointer. With the default comparator, a nil pointer and
+// empty struct are treated as different and so we would always treat the
+// CRD as not synced and would continually try and write it to Consul.
+func transparentProxyConfigComparer(a, b *capi.TransparentProxyConfig) bool {
+	empty := capi.TransparentProxyConfig{}
+	if a == nil && b == nil {
+		return true
+	}
```

Comment:
Should we add a comment saying why we need to compare when both are nil (I'm guessing is so that we can dereference it later)?

Reply:
I refactored this to just:

```go
	if a == nil && b != nil && *b == empty {
		return true
	}
	if b == nil && a != nil && *a == empty {
		return true
	}
```